src/src/PluginStructs/P123_data_struct.cpp

```diff
@@ -0,0 +1,707 @@
+      } else {
+        data = (data - 5) / 10;
+      }
+      data *= 10;
+    }
+    */
+
+    // save value
+    last_temp_val =  data / 100.0f;
+  }
+  
+  error|= readADC(i2caddr,filter_power);
+
+  return error;
+}
+
+
+
+int8_t P123_data_struct::requestADC(uint8_t i2caddr)
+{
+  
+  uint8_t reg;
+  int8_t error;
+
+  //set VOUT
+  // Get the current register value
+  error = readRegister(i2caddr, SI7013_READ_REG2, &reg);
+  if ( error == 0) {
+       
+      // Prepare to write to the register value
+    Wire.beginTransmission(i2caddr);
+    Wire.write(SI7013_WRITE_REG2);
+    
+    Wire.write(reg | (1+2+4+64) );//set last three bits to 1 (VIN bufered, Vref=VDD, VOUT=VDD) and No-Hold for bit 6
+
+    Wire.endTransmission();
+ 
+    //read adc
+    Wire.beginTransmission(i2caddr);
+    Wire.write(SI7013_READ_ADC);
+    Wire.endTransmission();
+
+    //delay(10); //wating for conversion to be done in the specs is mentioned 7ms in normal mode
+  }
+  return error;
+}
+
+int8_t P123_data_struct::readADC(uint8_t i2caddr, uint8_t filter_power)
+{
+  if ( Wire.requestFrom(i2caddr, 2u) < 2 ) {
+    return -1;
+  }
+
+  // Comes back in two bytes, data(MSB) / data(LSB) with no Checksum
+  uint16_t raw  = ((uint16_t) Wire.read()) << 8;
+  raw |= Wire.read();
+
+  //Calculate Moving average where 2^filter_power is the moving window of points
+  //MA*[i]= MA*[i-1] +X[i] - MA*[i-1]/N
+  last_adc_val = last_adc_val + raw - (last_adc_val>>filter_power);
+  
+  //set vout to gnd to not consume power
+  Wire.beginTransmission(i2caddr);
+    Wire.write(SI7013_WRITE_REG2);
+    Wire.write(SI7013_REG2_DEFAULT);
+  return (int8_t) Wire.endTransmission();
+}
+
+/* ======================================================================
+Function: Plugin_123_si7013_setResolution
+Purpose : Sets the sensor resolution to one of four levels
+Input   : see #define default is SI7013_RESOLUTION_14T_12RH
+Output  : 0 if okay
+Comments: -
+====================================================================== */
+int8_t P123_data_struct::setResolution(uint8_t i2caddr, uint8_t resolution)
+{
+  uint8_t reg;
+  uint8_t error;
+
+  // Get the current register value
+  error = readRegister(i2caddr, SI7013_READ_REG1, &reg);
+  if ( error == 0) {
+    // remove resolution bits
+    reg &= SI7013_RESOLUTION_MASK ;
+
+    // Prepare to write to the register value
+    Wire.beginTransmission(i2caddr);
+    Wire.write(SI7013_WRITE_REG1);
+
+    // Write the new resolution bits but clear unused before
+    Wire.write(reg | ( resolution &= ~SI7013_RESOLUTION_MASK) );
+    return (int8_t) Wire.endTransmission();
+  }
+
+  return error;
+}
+
+
+int8_t P123_data_struct::softReset(uint8_t i2caddr){
+  // Prepare to write to the register value
+    Wire.beginTransmission(i2caddr);
+    Wire.write(SI7013_SOFT_RESET);
+    Wire.endTransmission();
+    delay(50);
+    return 0;
+}
+
+
+int8_t P123_data_struct::readRevision(uint8_t i2caddr) {
+  Wire.beginTransmission(i2caddr);
```

Comment:
Can you please change these `Wire.xxx` calls to their matching functions in `src/Helpers/I2C_access.h` ?

The problem with repeating all these standard I2C calls is 2-fold:
- I always have to check if they obey the I2C standards, which isn't trivial.
- They add to the compiled bin size compared to single functions.

Reply:
working on this one.

Reply:
int32_t I2C_read24_reg(uint8_t i2caddr, uint8_t reg) does not checks for errors

Reply:
it seems that only the read 8bit is checking for errors and sets an ok flag.

Reply:
is I2C access used anywhere currently ?
Show references returns nothing.

Reply:
also the communication to the SI7013 is split in two separate parts:
first sending the command then later on reading the values.
This new API does not allows for this approach.
check the method P123_data_struct::update(..) which keeps track of states and handle the necessary timing.

The approach here is to start_measurement then later on check the data in another state of the sensor. 
The entire system behaves like a FSM allowing for complex activities to be tracked individually with no need of static delays.

Reply:
> is I2C access used anywhere currently ? Show references returns nothing.

BME280 plugin is using this code as it needed to read chunks of data in 1 I2C call.

Reply:
> The approach here is to start_measurement then later on check the data in another state of the sensor.
> The entire system behaves like a FSM allowing for complex activities to be tracked individually with no need of static delays.

That's how it should be, not using calls to `delay()`.

Reply:
> BME280 plugin is using this code as it needed to read chunks of data in 1 I2C call.

actually there is a problem there as all errors are hidden:
if (I2C_read8_reg(i2cAddress, BMx280_REGISTER_STATUS) & 0x08) {
    return false;
  }

Reply:
to do this async I need the 3 methods:
Start
CheckIfDataAvailable
GetData

currently the last two are combined together and the method signature are strange: 
usually status is the return of the function in C and any data to be set is passed as a pointer.

Reply:
I could deal with the last two combined if the return is clear when data is not available.

Reply:
OK, so you even want to split the I2C transfers into async steps.
Not sure if that will make a difference here as the I2C transfer itself is still blocking code.

Or maybe I don't understand what you're trying to do here?

Reply:
so the i2c_access methods do not allow for waiting in between the write command and read.
The datasheet requires certain timing in between, thus the required split.

1. se send a command to start the read of temp, hum, adc
2. wait a certain time based on datasheet (during this time another i2c device can be accessed)
3. request to read the data 
4. read the data as needed.

Reply:
That's what is also done on the BME280 as you initiate the start of the measurement and then have to wait for the measurement to complete.

Reply:
but that one still uses the wire lib not the new lib